Move error handling logic to the connection layer

lutovich opened the pull request:
It was previously spread between `RoutingSession`, `RoutingTransaction` and `RoutingStatementResult` and they all did the same thing when connection errors happened.

This PR introduces `RoutingPooledConnection` that handles errors for routing driver by removing entries from the routing table and purging connection pool for failed address.

lutovich committed:
Move error handling logic to the connection layer

It was previously spread between `RoutingSession`, `RoutingTransaction` and
`RoutingStatementResult` and they all did the same thing when connection
errors happened.

This commit introduces `RoutingPooledConnection` that handles errors for
routing driver by removing entries from the routing table and purging
connection pool for failed address.

zhenlineo commented on driver/src/main/java/org/neo4j/driver/internal/cluster/RoutingPooledConnection.java:
```diff
@@ -0,0 +1,289 @@
+            // In the future, we might be able to move this logic to the server.
+            switch ( accessMode )
+            {
+            case READ:
```
We actually do not need this `READ/WRITE` mode as the error itself already self explained:

`Neo.ClientError.Cluster.NotALeader` -> server at xx no longer accepts writes
`Neo.ClientError.General.ForbiddenOnReadOnlyDatabase` -> Write queries cannot be performed in `READ` access mode

So we could totally remove access mode from this class

lutovich replied:
It is possible to get `NotALeader` when trying to write in a read session with connection towards follower and `ForbiddenOnReadOnlyDatabase` when connection is towards read replica. So it looks like we still need this switch on the access mode.

zhenlineo commented on driver/src/main/java/org/neo4j/driver/internal/spi/PooledConnection.java:
```diff
@@ -0,0 +1,26 @@
+
+public interface PooledConnection extends Connection
+{
+    long lastUsedTimestamp();
```
We should also move 
`onError`
`hasUnrecoverableErrors` 
from `Connection` to `PooledConnection`

zhenlineo commented on driver/src/main/java/org/neo4j/driver/internal/cluster/RoutingPooledConnection.java:
```diff
@@ -0,0 +1,289 @@
+
+    private static boolean isFailureToWrite( ClientException e )
+    {
+        return e.code().equals( "Neo.ClientError.Cluster.NotALeader" ) ||
```
Warning: you might get `ClientException` without code field at this stage :)

zhenlineo reviewed (changes requested):
+1

lutovich committed:
Null-safe error code checks in connection layer

lutovich committed:
Move couple methods from Connection to PooledConnection

Methods `#onError()` and `#hasUnrecoverableErrors()` are only useful in pooled
connection and their implementations threw exceptions in regular connection.
This commit moves them to `PooledConnection` interface.

lutovich commented:
@zhenlineo comments should be addressed now
